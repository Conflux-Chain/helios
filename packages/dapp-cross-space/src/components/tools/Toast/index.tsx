import React, {memo} from 'react'
import {useSpring, a} from '@react-spring/web'
import {PopupClass} from '../../Popup'

const Toast = new PopupClass()
Toast.setListStyle({
  top: 'unset',
  left: 'unset',
  transform: 'unset',
  right: '12px',
  bottom: '6px',
  flexDirection: 'column-reverse',
})

const ToastComponent: React.FC<{text: string; duration: number}> = memo(
  ({text, duration}) => {
    const props = useSpring({
      from: {transform: 'translateX(-100%)'},
      to: {transform: 'translateX(0)'},
      config: {duration: 6000},
    })

    return (
      <div className="relative flex justify-center items-center min-w-[240px] h-[40px] px-[32px] bg-[#445159] text-white rounded overflow-hidden">
        {text}
        {duration ? (
          <a.div
            className="absolute bottom-0 w-full h-[4px] bg-gradient-to-l from-[#15C184] to-[#2959B4]"
            style={props}
          />
        ) : null}
      </div>
    )
  },
)

const showToast = (text: string, config?: any) =>
  Toast.show({
    Content: <ToastComponent text={text} duration={config?.duration ?? 6000} />,
    duration: 6000,
    ...config,
  })

export const hideToast = Toast.hide
export default showToast
