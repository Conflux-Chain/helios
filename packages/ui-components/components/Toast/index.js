import PropTypes from 'prop-types'
import React, {useEffect} from 'react'

function Toast({
  className = '',
  open,
  icon,
  title,
  content,
  autoHideDuration = 2000,
  type = 'block',
  onClose,
  ...props
}) {
  useEffect(() => {
    if (open) setTimeout(() => onClose && onClose(), autoHideDuration)
  }, [open, autoHideDuration, onClose])
  if (!open) return null
  const iconComp = icon
    ? React.cloneElement(icon, {
        className: `w-13 h-13 mb-2 ${icon.props.className || ''}`,
      })
    : null

  if (type === 'block') {
    return (
      <div
        className={`absolute flex flex-col items-center justify-center rounded-sm bg-black bg-opacity-60 p-3 w-35 max-h-52 shadow-1 z-20 ${className}`}
        {...props}
      >
        {iconComp}
        {title && (
          <div className="text-base text-white font-medium mb-0.5">{title}</div>
        )}
        <div className="text-xs text-gray-20">{content}</div>
      </div>
    )
  } else if (type === 'line' || !title) {
    return (
      <div
        className={`absolute flex items-center justify-center h-10 px-4 max-w-sm rounded-sm bg-black bg-opacity-60 shadow-1 z-20 ${className}`}
        {...props}
      >
        {icon && <img src={icon} alt="status" className="mr-2 w-6 h-6" />}
        <div className="text-xs text-white">{content}</div>
      </div>
    )
  }
  return null
}

Toast.propTypes = {
  className: PropTypes.string,
  open: PropTypes.bool,
  icon: PropTypes.node,
  title: PropTypes.string,
  content: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  autoHideDuration: PropTypes.number,
  type: PropTypes.oneOf(['block', 'line']),
  onClose: PropTypes.func,
  top: PropTypes.string,
}

export default Toast
