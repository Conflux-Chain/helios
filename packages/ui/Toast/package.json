{
  "name": "@cfxjs/toast",
  "type": "module",
  "main": "index.js"
}
