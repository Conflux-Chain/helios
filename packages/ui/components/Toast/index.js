import React, {useEffect} from 'react'
import PropTypes from 'prop-types'

function Toast({
  className = '',
  open,
  icon,
  title,
  content,
  autoHideDuration = 2000,
  type = 'block',
  onClose,
  ...props
}) {
  useEffect(() => {
    if (open) setTimeout(() => onClose && onClose(), autoHideDuration)
  }, [open, autoHideDuration, onClose])
  if (!open) return null
  const iconComp = icon
    ? React.cloneElement(icon, {
        className: `w-13 h-13 mb-2 ${icon.props.className || ''}`,
      })
    : null

  if (type === 'block') {
    return (
      <div
        data-testid="block-toast-wrapper"
        className={`absolute flex flex-col items-center justify-center rounded-sm bg-[#000] bg-opacity-40 p-3 w-35 max-h-52 shadow-1 z-20 ${className}`}
        {...props}
      >
        {iconComp}
        {title && (
          <div className="text-base text-white font-medium mb-0.5">{title}</div>
        )}
        <div className="text-xs text-gray-20">{content}</div>
      </div>
    )
  } else if (type === 'line' || !title) {
    return (
      <div
        data-testid="line-toast-wrapper"
        className={`absolute flex items-center justify-center h-6 px-2 max-w-sm rounded-sm bg-[#000] bg-opacity-40 z-20 ${className}`}
        {...props}
      >
        {icon && <img src={icon} alt="status" className="mr-0.5 w-3 h-3" />}
        <div className="text-xs text-white whitespace-nowrap">{content}</div>
      </div>
    )
  }
  return null
}

Toast.propTypes = {
  className: PropTypes.string,
  open: PropTypes.bool,
  icon: PropTypes.node,
  title: PropTypes.string,
  content: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  autoHideDuration: PropTypes.number,
  type: PropTypes.oneOf(['block', 'line']),
  onClose: PropTypes.func,
  top: PropTypes.string,
}

export default Toast
