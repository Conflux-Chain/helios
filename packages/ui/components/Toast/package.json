{
  "private": true,
  "version": "0.0.0",
  "name": "@fluent-wallet/component-toast",
  "type": "module",
  "main": "index.js",
  "peerDependencies": {
    "react": "^17",
    "react-dom": "^17"
  }
}
