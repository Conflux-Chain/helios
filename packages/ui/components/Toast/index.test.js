import {render, screen, cleanup} from '@testing-library/react'
import {describe, expect, vi, beforeEach, afterEach, it} from 'vitest'
import Toast from './index.js'

beforeEach(cleanup)
describe('Tag', () => {
  beforeEach(() => {
    vi.useFakeTimers()
  })
  afterEach(() => {
    vi.useRealTimers()
  })
  it('test snapshot', () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const snapshot = render(<Toast type="block" open={true} />)
    expect(snapshot).toMatchSnapshot()
  })
  it('test block type', () => {
    render(<Toast type="block" open={true} />)
    expect(screen.getByTestId('block-toast-wrapper')).toBeInTheDocument()
  })
  it('test line type', () => {
    render(<Toast type="line" open={true} />)
    expect(screen.getByTestId('line-toast-wrapper')).toBeInTheDocument()
  })
  it('test default type', () => {
    render(<Toast open={true} />)
    expect(screen.getByTestId('block-toast-wrapper')).toBeInTheDocument()
  })
  it('test open is false', () => {
    render(<Toast />)
    expect(screen.queryByTestId('block-toast-wrapper')).not.toBeInTheDocument()
  })
  it('test open is true', () => {
    render(<Toast open={true} />)
    expect(screen.getByTestId('block-toast-wrapper')).toBeInTheDocument()
  })
  it('test class name', () => {
    render(<Toast className="test-class-name" open={true} />)
    expect(screen.getByTestId('block-toast-wrapper')).toHaveClass(
      'test-class-name',
    )
  })
  it('test icon', () => {
    render(<Toast icon={<div data-testid="icon" />} open={true} />)
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('test title', () => {
    render(<Toast type="block" open={true} title="test-title" />)
    expect(screen.getByText('test-title')).toBeInTheDocument()
  })
  it('test content', () => {
    render(<Toast open={true} content="test-content" />)
    expect(screen.getByText('test-content')).toBeInTheDocument()
  })
  it('test onClose', async () => {
    const onClose = vi.fn()
    render(<Toast open={true} content="test-content" onClose={onClose} />)
    vi.advanceTimersByTime(2000)
    expect(onClose).toHaveBeenCalledTimes(1)
  })
  it('test autoHideDuration', async () => {
    const onClose = vi.fn()
    render(
      <Toast
        open={true}
        content="test-content"
        onClose={onClose}
        autoHideDuration={3000}
      />,
    )
    vi.advanceTimersByTime(3000)
    expect(onClose).toHaveBeenCalledTimes(1)
  })
})
