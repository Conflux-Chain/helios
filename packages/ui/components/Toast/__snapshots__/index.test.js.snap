// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag test snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="absolute flex flex-col items-center justify-center rounded-sm bg-[#000] bg-opacity-40 p-3 w-35 max-h-52 shadow-1 z-20 "
        data-testid="block-toast-wrapper"
      >
        <div
          class="text-xs text-gray-20"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="absolute flex flex-col items-center justify-center rounded-sm bg-[#000] bg-opacity-40 p-3 w-35 max-h-52 shadow-1 z-20 "
      data-testid="block-toast-wrapper"
    >
      <div
        class="text-xs text-gray-20"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
